modulemd/modulemd-module.c

```diff
@@ -916,30 +858,25 @@ modulemd_module_add_obsoletes (ModulemdModule *self,
             }
         }
 
-      switch (modulemd_module_stream_get_mdversion (stream))
+      if (modulemd_module_stream_get_mdversion (stream) <
+          MD_MODULESTREAM_VERSION_TWO)
         {
-        case MD_MODULESTREAM_VERSION_TWO:
-          current_obsoletes = modulemd_module_stream_v2_get_obsoletes (
-            (ModulemdModuleStreamV2 *)stream);
-          break;
-
-        case MD_MODULESTREAM_VERSION_THREE:
-          current_obsoletes = modulemd_module_stream_v3_get_obsoletes (
-            (ModulemdModuleStreamV3 *)stream);
-          break;
-
-        default:
-          g_info (
-            "ModuleStream obsoletes requires mdversion two or greater: "
-            "module: %s, stream: %s, context: %s, mdversion: %u",
-            modulemd_module_get_module_name (self),
-            stream_str,
-            context_str,
-            (uint)modulemd_module_stream_get_mdversion (stream));
-          g_return_if_reached ();
+          /* If the stream we need to associate obsoletes is of a lower version than 2
+           * upgrade it to the version 2, so it can use obsoletes.
+           *
+           * We only call this if the mdversion is definitely lower, because the
+           * upgrade() routine is not designed to handle downgrades.
+           */
+          ModulemdModuleStream *newstream = NULL;
+          newstream = modulemd_module_stream_upgrade (
+            stream, MD_MODULESTREAM_VERSION_TWO, NULL);
+          g_ptr_array_remove (self->streams, stream);
+          g_ptr_array_add (self->streams, newstream);
```

Comment:
`g_ptr_array_index(foo, i)` is just a macro that expands to `foo->pdata[i]`, so that example, post-processed, is
```
mc->measurements->pdata[framenumber] = cpy;
```